Fix for AOTI + CUDAGraphs when calling from Python

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148601

**Background**: I've been comparing performance of torch.compile vs. torch.export + AOTI (specifically, loaded from Python) on the Flux model and found a ~1.4% performance decrease with the latter. The trace shows that CUDAGraphs are not utilized for torch.export + AOTI, leading to higher overhead.

When trying to manually CUDAGraph the loaded, previously exported + AOTIed model (thanks to @eellison for the logic here), I get:
```
Error: operation not permitted when stream is capturing
```

@desertfire confirms that this is due to multi-threading logic on the AOTI runtime side (in `AOTIModelContainer` / `AOTIModel`) conflicting with the use of CUDAGraphs.

**Fix**: This PR takes the approach of providing an alternate, single-threaded method for running loaded models with the AOTI runtime. Details:
* Python side introduces a new flag to enable this behavior (needs a better name): `torch._inductor.package.load_package(..., run_single_threaded=False)`
    * This flag is passed down to the C++ side's `AOTIModelPackageLoader`, which passes it to the `CreateAOTIModelRunnerFunc` during `AOTIModelContainerRunner` construction.
* C++ side introduces single-threaded alternatives to model running and model container running:
    * `AOTIModelContainer.run_single_threaded()` / `AOTIModel.run_single_threaded()`. The interfaces match those of `run()`, but the synchronization logic has been removed.
    * Introduces `AOTInductorModelContainerRunSingleThreaded` to AOTI's `interface.h`; this is invoked by the `AOTIModelContainerRunner` utility class when `run_single_threaded=true`.

I've verified on both a small repro and my real-world use case that I can manually CUDAGraph a loaded model that was previously exported + AOTIed.

**Future work:**
* Flip default value to `run_single_threaded=True` as Python-side inference doesn't take advantage of the AOTI runtime thread pool
    * There are some BC concerns here - models need to be re-serialized so the .so contains the new `AOTInductorModelContainerRunSingleThreaded` interface func. We can flip the default value and warn (instead of crashing) if the `AOTInductorModelContainerRunSingleThreaded` symbol does not exist.
* Compose with cudagraph trees as opposed to manual cuda graph wrapping

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

jbschlosser committed:
Fix for AOTI + CUDAGraphs when calling from Python

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix for AOTI + CUDAGraphs when calling from Python"


TODO: tests. I'm currently verifying this on my real-world Flux use case but I'll distill this down to a nice test afterwards.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_package/model_package_loader.h:
```diff
@@ -10,7 +10,8 @@ class TORCH_API AOTIModelPackageLoader {
   AOTIModelPackageLoader(const std::string& model_package_path);
```
Better to update this API as well.

desertfire commented:
Let me know what the new perf looks like.

jbschlosser commented:
> Let me know what the new perf looks like.

looks great! I'm no longer seeing a substantial difference between compile vs. export + AOTI + cudagraphs, where previously I was seeing a ~1.4% decrease from the export variant.

Compile:

![compile_perf](https://github.com/user-attachments/assets/4f0ca6d3-c07e-4d0c-9d46-dad7daecb2de)

Export + AOTI + manual CUDAGraphs:

![export_perf](https://github.com/user-attachments/assets/6e807188-8b3e-48f6-950f-2ec7efd0351b)

jbschlosser commented on torch/_inductor/package/package.py:
```diff
@@ -274,7 +274,9 @@ def __deepcopy__(self, memo: Optional[dict[Any, Any]]) -> "AOTICompiledModel":
 
-def load_package(path: FileLike, model_name: str = "model") -> AOTICompiledModel:  # type: ignore[type-arg]
+def load_package(
+    path: FileLike, model_name: str = "model", run_single_threaded: bool = False
```
sounds good, I'll flip this and the one in `torch/csrc/inductor/aoti_eager/kernel_holder.cpp` in a follow-up.

eellison reviewed (commented):
A nice follow up would be for the aoti model to compose with cudagraph trees, as opposed to manual cuda graph wrapping. this would add the nice benefits of memory pool sharing, dynamic shape handling, etc, etc. although for only wrapping a single model it's not really essential. 

I think it should (pretty much) just work, but there might some metadata to maintain that we would need to thread to cudagraphify.

henrylhtsang commented:
Question:
1. how to use? Just run_single_threaded=True? do we need to enable cuda graphs?
2. is there a diff version to try out in fbcode?

jbschlosser committed:
Update on "Fix for AOTI + CUDAGraphs when calling from Python"


**Background**: I've been comparing performance of torch.compile vs. torch.export + AOTI (specifically, loaded from Python) on the Flux model and found a ~1.4% performance decrease with the latter. The trace shows that CUDAGraphs are not utilized for torch.export + AOTI, leading to higher overhead.

When trying to manually CUDAGraph the loaded, previously exported + AOTIed model (thanks to eellison for the logic here), I get:
```
Error: operation not permitted when stream is capturing
```

desertfire confirms that this is due to multi-threading logic on the AOTI runtime side (in `AOTIModelContainer` / `AOTIModel`) conflicting with the use of CUDAGraphs.

**Fix**: This PR takes the approach of providing an alternate, single-threaded method for running loaded models with the AOTI runtime. Details:
* Python side introduces a new flag to enable this behavior (needs a better name): `torch._inductor.package.load_package(..., run_single_threaded=False)`
    * This flag is passed down to the C++ side's `AOTIModelPackageLoader`, which passes it to the `CreateAOTIModelRunnerFunc` during `AOTIModelContainerRunner` construction.
* C++ side introduces single-threaded alternatives to model running and model container running:
    * `AOTIModelContainer.run_single_threaded()` / `AOTIModel.run_single_threaded()`. The interfaces match those of `run()`, but the synchronization logic has been removed.
    * Introduces `AOTInductorModelContainerRunSingleThreaded` to AOTI's `interface.h`; this is invoked by the `AOTIModelContainerRunner` utility class when `run_single_threaded=true`.

I've verified on both a small repro and my real-world use case that I can manually CUDAGraph a loaded model that was previously exported + AOTIed.

**TODO**: Test case in progress.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jbschlosser commented:
@henrylhtsang 

> how to use? Just run_single_threaded=True? do we need to enable cuda graphs?

I just uploaded a test case in `test/inductor/test_aot_inductor.py` that should give you an idea on how to use it. The short answer is use `torch._inductor.aoti_load_package(..., run_single_threaded=True)` and manually capture a CUDAGraph for running the loaded model. The test case includes a simplistic `cudagraph()` function wrapper that copies inputs into fixed memory locations for the graph and clones outputs on exit.

> is there a diff version to try out in fbcode?

not yet but soon this will be imported - you could try to `ghimport` it if you want it faster

jbschlosser commented:
@desertfire one issue I just realized wrt flipping the default value is that this is not backwards compatible with previously serialized models. This functionality requires re-serializing so that the new interface function `AOTInductorModelContainerRunSingleThreaded` will be available in the .so.

desertfire commented:
> @desertfire one issue I just realized wrt flipping the default value is that this is not backwards compatible with previously serialized models. This functionality requires re-serializing so that the new interface function `AOTInductorModelContainerRunSingleThreaded` will be available in the .so.

Yes, you are right. I thought it's less an issue since AOTI+python is not a major use case, but since there are already use cases like it out there, we shouldn't break them.

As a future work, I think we can still flip the default value, and check if `AOTInductorModelContainerRunSingleThreaded` symbol exist. If not, we could print a warning instead of crash it.

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -4553,6 +4553,72 @@ def forward(self, x):
+        # warmup -> run with CUDAGraphs
+        for _ in range(3):
+            optimized(**model_kwargs)
+
```
Let's add a result comparison with eager.

jbschlosser committed:
Update on "Fix for AOTI + CUDAGraphs when calling from Python"


**Background**: I've been comparing performance of torch.compile vs. torch.export + AOTI (specifically, loaded from Python) on the Flux model and found a ~1.4% performance decrease with the latter. The trace shows that CUDAGraphs are not utilized for torch.export + AOTI, leading to higher overhead.

When trying to manually CUDAGraph the loaded, previously exported + AOTIed model (thanks to eellison for the logic here), I get:
```
Error: operation not permitted when stream is capturing
```

desertfire confirms that this is due to multi-threading logic on the AOTI runtime side (in `AOTIModelContainer` / `AOTIModel`) conflicting with the use of CUDAGraphs.

**Fix**: This PR takes the approach of providing an alternate, single-threaded method for running loaded models with the AOTI runtime. Details:
* Python side introduces a new flag to enable this behavior (needs a better name): `torch._inductor.package.load_package(..., run_single_threaded=False)`
    * This flag is passed down to the C++ side's `AOTIModelPackageLoader`, which passes it to the `CreateAOTIModelRunnerFunc` during `AOTIModelContainerRunner` construction.
* C++ side introduces single-threaded alternatives to model running and model container running:
    * `AOTIModelContainer.run_single_threaded()` / `AOTIModel.run_single_threaded()`. The interfaces match those of `run()`, but the synchronization logic has been removed.
    * Introduces `AOTInductorModelContainerRunSingleThreaded` to AOTI's `interface.h`; this is invoked by the `AOTIModelContainerRunner` utility class when `run_single_threaded=true`.

I've verified on both a small repro and my real-world use case that I can manually CUDAGraph a loaded model that was previously exported + AOTIed.

**Future work:**
* Flip default value to `run_single_threaded=True` as Python-side inference doesn't take advantage of the AOTI runtime thread pool
    * There are some BC concerns here - models need to be re-serialized so the .so contains the new `AOTInductorModelContainerRunSingleThreaded` interface func. We can flip the default value and warn (instead of crashing) if the `AOTInductorModelContainerRunSingleThreaded` symbol does not exist.
* Compose with cudagraph trees as opposed to manual cuda graph wrapping

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

henrylhtsang commented:
noob question: Any idea if we need to pay attention to stream at all?

desertfire commented:
> noob question: Any idea if we need to pay attention to stream at all?

Are you asking how to use stream from the Python interface?
